Factory updates and unification

tomeichlersmith opened the pull request:
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
This resolves #1465 and resolves #1698 

Besides fixing the issue with the SimCore/Factory I also did some refactoring to simplify/unify the dynamic factory code.
- Move SimCore/Factory into Framework and remove Framework/PluginFactory (and SimCore/Geo/ParserFactory)
- Refactor EventProcessor/Produer/Analyzer to function with templated Factory
- New registration macros do not require separating namespace and class name so I've removed the `*_NS` macros
- When I was updating the Factory documentation, I noticed that `doxygen` was running over all the files in the `build` subdirectory as well, so I am also updating the doxyfile so that it properly excludes the generated/copied files.
- The Factory does not throw its own exceptions, instead it returns `nullptr` if it is unable to create an object by name leaving it up to the caller to decide how to handle the unable-to-create situation (and what to write into the exception). I chose to do this because some exceptions need to have more detail than others.

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments.
- [x] I read, understood and follow the [coding rules](https://ldmx-software.github.io/developing/coding-rules.html).
- [x] I ran my developments and the following shows that they are successful.
- [x] I have added `just test` into the clang-lto workflow which includes running the basic simulation. Running the basic simulation fails on `trunk` as documented in #1465 but it doesn't anymore.
- [x] I manually ran a asan-ubsan build locally to see if I introduced any new memory leaks and I found none
- [x] Look at how the new error message appears in terminal
- [x] Make sure single-file processors still functions. I noticed when testing this that the single-file compilation complains about the C++ standard so I have added the C++20 standard to the single-file compilation command `-std=c++20`.
- [x] Undo library/file movement in SimCore? I don't think its necessary and I'm unsure if its even helpful from a code-understandability perspective. _Reply_: I'm going to leave it as is. I don't think reverting it makes any more sense anyways.

tomeichlersmith committed:
move library dyn loading into Process where it is called

now Framework/PluginFactory and SimCore/Factory have the same
responsibilities and so we should (hopefully) be able to replace one
with the other

tomeichlersmith committed:
add enable_exports property

tomeichlersmith committed:
dont compile unrelated stuff

just tring to make testing faster

tomeichlersmith committed:
change justfile to be able to have parallel builds

don't plan to merge this, just helping myself for testing

tomeichlersmith committed:
add status messages and a static counter inside the class

tomeichlersmith committed:
try manually loading the SDs library

tomeichlersmith committed:
compiling libraries as *dependencies* of SimCore functions

This only works for Clang and not for GCC.
For Clang, I _think_ this works because it uniquely defines a single source of
truth of the static variable for the compiler.
For GCC, we get an "undefined symbol error" when attempting to load the
PhotoNuclearModels library. The symbol is the implementation of the
*base* class which resides in SimCore and not SimCore_PhotoNuclearModels
so I'm going to try moving the base class into the specific library now.

tomeichlersmith committed:
move PhotoNuclearModel base class into PhotoNuclearModels library

tomeichlersmith committed:
move SensitiveDetector base class into SDs library

tomeichlersmith committed:
move PrimaryGenerator base class into Generators library

tomeichlersmith committed:
remove previous manual loading

tomeichlersmith committed:
move XsecBiasingOperator base class into BiasOperators library

tomeichlersmith committed:
re-enable all parts of ldmx-sw for further testing

tomeichlersmith committed:
add ecal_pn test for Biasing UserActions, clang compilation

tomeichlersmith committed:
move Factory singleton instance into Prototype class

This allows there to be exactly one TU where the `static` variable is
defined. Before, when the `static Factory` was in the header with the
templated class itself, I believe it was being erroneously _defined_ in
multiple TUs. Now that it is only ever defined in one TU, the dynamic
factory appears to be working with both GCC and Clang with LTO.

tomeichlersmith committed:
Revert "add enable_exports property"

This reverts commit 03cec417430dc4e1fe6f0ad3a400b9ee17ed100b.

This additional property is not necessary and I think removing it helps
make the CMake code a bit more understandable. Trying to keep things
clean.

tomeichlersmith committed:
mv SimCore/Factory to Framework/Factory

tomeichlersmith committed:
SHOULD REVERT: shortening test configs since I'm just testing loading

tomeichlersmith committed:
replace PluginFactory with templated Factory

tomeichlersmith committed:
move registration macro definition to Factory level

remove ParserFactory now that it is redundant, replace with framework::Factory

tomeichlersmith committed:
better exclude rules for doxygen

tomeichlersmith committed:
rename macros to align with our coding rules/habits

tomeichlersmith committed:
re-enable testing

tomeichlersmith committed:
clang-format after all these changes

tomeichlersmith committed:
update links and docs of Factory

tomeichlersmith committed:
revert developing changes to test configs and justfile

tomeichlersmith committed:
drop one level of macro indirection

The newer FACTORY_REGISTRATION macro does not need to separate the
namespace from the class name in order to function, so I am dropping the
*_NS macros from Framework and dropping the NS argument from the
UserAction macro.

This was simple enough syntax to rewrite with sed which made it easy
enough for me to do.

  sed '/^DECLARE_/ s/_NS//; /^DECLARE_/ s/,\s*/::/'

Only applying this to files where I knew a declaration took place.

tomeichlersmith committed:
should be able to run tests with clang-lto build now

tomeichlersmith committed:
add more detailed things to check to UnableToCreate error

tomeichlersmith committed:
add C++20 standard flag to from-file compilation command

github-actions[bot] committed:
Apply clang-format

tomeichlersmith committed:
add extra space for readability

tomeichlersmith committed:
update documentation comment on Analyzer::beforeNewRun

tvami reviewed (approved):
This is awesome work, thanks Tom!

tomeichlersmith committed:
cleanup SimCore/CMakeLists.txt since I'm already here

isolate SimCore::Geo and remove unnecessary registered_targets link

EinarElen commented:
+1 on Tamas.

Regarding returning a nullptr, could we not return a std::optional? Nullptrs have a habit of being ignored (std::expected would be better but that's c++23 😔)

EinarElen commented on Framework/include/Framework/ConditionsObjectProvider.h:
```diff
@@ -72,7 +70,7 @@ class ConditionsObjectProvider {
    * Class destructor.
    */
-  virtual ~ConditionsObjectProvider() { ; }
+  virtual ~ConditionsObjectProvider() = default;
```
Underappreciated patch in the middle of all of this

EinarElen commented on Framework/include/Framework/EventProcessor.h:
```diff
@@ -312,6 +310,20 @@ class Analyzer : public EventProcessor {
+   * that any derived Analyzer will not be able to compile an
+   * override of the beforeNewRun callback.
+   */
+  virtual void beforeNewRun(ldmx::RunHeader &runHeader) final {}
```
Really nice

EinarElen commented on Framework/src/Framework/Process.cxx:
```diff
@@ -80,14 +93,20 @@ Process::Process(const framework::config::Parameters &configuration)
+    EventProcessor *ep =
+        EventProcessor::Factory::get().make(className, instanceName, *this);
+    if (not ep) {
+      EXCEPTION_RAISE("UnableToCreate",
+                      "The EventProcessor Factory was unable to create " +
+                          instanceName + " of type " + className +
+                          ". Did you inherit from framework::Producer or "
+                          "framework::Analyzer? "
+                          "Did you DECLARE_PRODUCER or DECLARE_ANALYZER in the "
+                          "implementation (.cxx) file? "
+                          "Did you use the class's full name (including "
+                          "namespaces) in the Python configuration class? "
+                          "Does the Python configuration class reference the "
+                          "correct library it is a part of?");
```
🔥

tomeichlersmith commented:
The reason I did not use `std::optional` is because then the returned object would have three possible states.

1. `std::nullopt`
2. `std::optional::has_value` is `true` but the pointer is `nullptr`
3. `std::optional::has_value` is `true` and the pointer is valid

I didn't like the confusion of having two different ways of reflecting that the returned pointer is invalid. I also don't like having to trust that we check for `nullptr`, but I think (hope?) that our additional warnings and the fact that the Factory code is not interacted with much will keep it from being an issue.

EinarElen commented:
Yeah if we were to return optional we would have to return a value type or make the API promise that optional.has_value() for this part always implies a valid pointer

tomeichlersmith commented:
Alright, I'm convinced. I was just looking into SimCore a bit more and found a few places where I forgot to check if the returned value was non-null. I'm going to change the return type of `Factory::make` to `std::optional<PrototypePtr>` and mark it as `[[nodiscard]]`.

tomeichlersmith committed:
update Factory to return [[nodiscard]] std::optional

This helps us make sure that we are checking that the factory
successfully created the object we asked for.
Additionally, I separated out the warehouse_ into its own derived
Factory (FactoryWithWarehouse) that can be used simply by using a
different macro. This is nice since many of the prototypes that use the
factory already have a central storage location for the objects and
don't need the warehouse carrying around copies of the pointers itself.

github-actions[bot] committed:
Apply clang-format
